staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,379 @@
+			return r.SubIndexValidator(invocation, index, schema, rootSchema, root, formats, opts...)
+		}
+	}
+}
+
+func (r *ratchetingValueValidator) Applies(value interface{}, kind reflect.Kind) bool {
+	return true
+}
+
+func (s *ratchetingValueValidator) SetPath(path string) {
+	s.schemaArgs.path = path
+}
+
+// Validate validates the value.
+//
+// During evaluation, a temporary tree of scratch storage is built for all
+// traversed field paths. It is necessary to build the tree to take advantage of
+// DeepEqual checks performed by lower levels of the object during validation without
+// greatly modifying `kube-openapi`'s implementation.
+//
+// The tree, and all cache storage/scratch space for the validation of a single
+// `value` passed to `Validate` are stored in the `invocation` variable and
+// thrown away at the end of the top-level call to `Validate`.
+//
+// `Validate` will create a node in the tree to represent this call to `Validate(value)`.
+// The node's main purpose is to store a lazily computed DeepEqual check between
+// the oldValue and the currently passed value. If the check is performed, it
+// will be stored in the node to be re-used by a parent node during a DeepEqual
+// comparison, if necessarily.
+//
+// It is assumed that for `ratchetingValueValidator` instances whose
+// `parentInvocation` is non-nil, that `Validate` will only be called on a
+// single value - the value of the parent's field or index
+//
+// This call has a side-effect of writing to it's parentInvocation a
+// lazily-computed equality.
+func (r *ratchetingValueValidator) Validate(value interface{}) *validate.Result {
+	// Create a new node in the validation scratch space tree
+	invocation := &ratchetInvocationScratch{
+		value:            value,
+		thunk:            r,
+		children:         map[interface{}]*ratchetInvocationScratch{},
+		comparisonResult: nil,
+	}
+
+	if r.parentInvocation != nil {
+		// If this is not a top-level node, attach it to the parent
+		// This is only correct to do if precondition that `Validate`
+		// is only used on a single value is being met.
+		r.parentInvocation.children[r.keyInParent] = invocation
+	}
+
+	opts := append([]validate.Option{
+		r.getValidateOptionFor(invocation),
+	}, r.options...)
+
+	s := validate.NewSchemaValidator(r.schema, r.root, r.path, r.knownFormats, opts...)
+
+	res := s.Validate(value)
+
+	if res.IsValid() {
+		return res
+	}
+
+	// Current ratcheting rule is to ratchet errors if the DeepEqual(old, new)
+	// is true.
+	if invocation.DeepEqualIfNecessary() {
+		newRes := &validate.Result{}
+		newRes.MergeAsWarnings(res)
+		return newRes
+	}
+
+	return res
+}
+
+func (r *ratchetingValueValidator) SubPropertyValidator(invocation *ratchetInvocationScratch, field string, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, options ...validate.Option) validate.ValueValidator {
```

Comment:
// SubPropertyValidator overrides the standard validator constructor for sub-properties by
// returning our special ratcheting variant.

Reply:
similar for the index variant below.

Reply:
added godoc for this function and index variant